python/tchannel/tornado/hyperbahn.py

```diff
@@ -22,18 +22,98 @@
 
 import json
 import logging
+import random
 
 import tornado.gen
 import tornado.ioloop
 
-from ..messages.error import ErrorCode
+from tchannel.tornado.timeout import timeout
 
-DEFAULT_TTL = 60  # seconds
+from ..errors import ProtocolError
+from ..errors import TimeoutError
+from .response import StatusCode
 
+DEFAULT_EXPO_BASE = 1.4  # try this first
+DEFAULT_MAX_DELAY = 10  # sec
+DEFAULT_MAX_ATTEMPT = 7  # pow(1.4, 8) > 10
+DEFAULT_DELAY = 60  # sec delay time for successful advertise
+DEFAULT_FIRST_ADVERTISE_TIME = 30  # sec
 
 log = logging.getLogger('tchannel')
 
 
+def _advertise(tchannel, service):
+    return tchannel.request(service='hyperbahn').send(
+        arg1='ad',  # advertise
+        arg2='',
+        arg3=json.dumps({
+            'services': [
+                {
+                    'serviceName': service,
+                    'cost': 0,
+                }
+            ]
+        }),
+        headers={'as': 'json'},
+        attempt_times=1,
+    )
+
+
+def _prepare_next_ad(attempt_counter):
+    if attempt_counter:
+        delay_time = random.uniform(
+            0, min(DEFAULT_MAX_DELAY, DEFAULT_EXPO_BASE ** attempt_counter)
+        )
+    else:
+        delay_time = DEFAULT_DELAY
+    attempt_counter = min(attempt_counter, DEFAULT_MAX_ATTEMPT)
+    return (attempt_counter, delay_time)
+
+
+@tornado.gen.coroutine
+def _regular_advertise(tchannel, service, attempt_counter=0):
```

Comment:
There is one different place. In first_advertise, if we get successful response we will return the response. It makes hard to extract common piece of code. There are exception, response not ok, and response ok to compare. Making a common method doesn't gain any thing